libs/src/services/discoveryProvider/DiscoveryProviderSelection.js

```diff
@@ -0,0 +1,187 @@
+        const isExpired = (Date.now() - timestamp) > DISCOVERY_PROVIDER_RESELECT_TIMEOUT
+        if (!inWhitelist || isExpired) {
+          this.clearCached()
+        } else {
+          return latestEndpoint
+        }
+      }
+    }
+    return null
+  }
+
+  /** Clears any cached discovery provider from localstorage */
+  clearCached () {
+    if (localStorage) {
+      localStorage.removeItem(DISCOVERY_PROVIDER_TIMESTAMP)
+    }
+  }
+
+  /** Sets a cached discovery provider in localstorage */
+  setCached (endpoint) {
+    localStorage.setItem(DISCOVERY_PROVIDER_TIMESTAMP, JSON.stringify({ endpoint, timestamp: Date.now() }))
+  }
+
+  /** Allows the selection take a shortcut if there's a cached provider */
+  shortcircuit () {
+    return this.getCached()
+  }
+
+  async select () {
+    const endpoint = await super.select()
+    this.setCached(endpoint)
+    return endpoint
+  }
+
+  /**
+   * Checks whether a given response is healthy:
+   * - Not behind in blocks
+   * - 200 response
+   * - Current version
+   *
+   * Other responses are collected in `this.backups` if
+   * - Behind by only a patch version
+   *
+   * @param {Response} response axios response
+   * @param {{ [key: string]: string}} urlMap health check urls mapped to their cannonical url
+   * e.g. https://discoveryprovider.audius.co/health_check => https://discoveryprovider.audius.co
+   */
+  isHealthy (response, urlMap) {
+    const { status, data } = response
+    const { block_difference: blockDiff, service, version } = data
+    if (status !== 200) return false
+    if (service !== DISCOVERY_SERVICE_NAME) return false
+    if (!semver.valid(version)) return false
+    if (!this.ethContracts.isValidSPVersion(version, this.currentVersion)) return false
+
+    if (
+      blockDiff > UNHEALTHY_BLOCK_DIFF ||
+      version !== this.currentVersion
+    ) {
+      this.addBackup(urlMap[response.config.url], response.data)
+      return false
+    }
+
+    return true
+  }
+
+  /**
+   * Estabilishes that connection to discovery providers has regressed
+   */
+  enterRegressedMode () {
+    console.info('Entering regressed mode')
+    this._regressedMode = true
+    setTimeout(() => {
+      console.info('Leaving regressed mode')
+      this._regressedMode = false
+    }, REGRESSED_MODE_TIMEOUT)
+  }
+
+  isInRegressedMode () {
+    return this._regressedMode
+  }
+
+  /**
+   * In the case of no "healthy" services, we resort to backups.
+   * 1. Pick the most recent version that's not behind
+   * 2. Pick the least behind provider and enter "regressed mode"
+   */
+  selectFromBackups () {
+    const versions = []
+    const blockDiffs = []
+
+    const versionMap = {}
+    const blockDiffMap = {}
+
+    // Go through each backup and record version and block diff maps
+    Object.keys(this.backups).forEach(backup => {
+      const { block_difference: blockDiff, version } = this.backups[backup]
+      versions.push(version)
```

Comment:
how does this work if a service has an older but invalid version endpoint? like if valid versions are 0.4, 0.6 and a backup endpoint is returning 0.5.x, would it get selected?

Reply:
seems like this is why we don't query past versions from chain - fwiw i don't think this is a blocker, but something to consider

Reply:
Yeah I didn't add a check for this. Good call. We definitely shouldn't be selecting 0.5.x

It would be really great if we had a versionFactoryClient.isValidVersion(version) rather than traversing backwards with query Index. do you know how hard that would be to add?

as a stop gap, I added code in the selectfrombackups that will go backwards through the versions. I don't want to fall into a scenario where we just keep trying to go backwards forever looking for discprovs with a specific version -- would be much easier just to look at the verion of a registered service and validate if we're allowed to take it. The former gets pretty costly because of all the contract method calls.

Reply:
https://github.com/AudiusProject/audius-protocol/pull/402/commits/b9e5a5016dc10ec2bba54ba995c6ed9f9820b7b2